### PR TITLE
Project backend rewrite

### PR DESCRIPTION
**Pull request by Prince781**
This work does a few things:
- Abstract the project backend system, making it much easier to write new backends like CMake or Autotools
- Improve Meson support to handle any kind of build target, not just a Vala compilation. This is especially useful for projects that use arbitrary commands to generate VAPI files (Geary for example).
- Improve the API for many things, from querying files, to handling paths in a platform-agnostic way, to making it easy to reconfigure an entire project if the build files are changed
- Miscellaneous improvements

### CODE REVIEW

**Comment by Prince781** [comment]
I need to rebase this, since the last commit is the only relevant part of this PR. However I'll do that after some of the other PRs are merged.

**Comment by Prince781** [comment]
So, the basic summary of this MR is that it greatly improves how projects are handled.

- Build targets are handled much better: first we create build tasks for any type of target, then we analyze dependence relationships between the targets, and finally we discard those targets that aren't necessary for Vala targets. The result is that projects like Gitg now only have a few errors instead of over a thousand
- When no backend is found, we now treat every file as a separate project and only create new projects when new files are opened. This behavior makes more sense since it's most likely the case that the user is opening a single file. If we detect `CMakeLists.txt` or `autogen.sh`, we notify them that those backends aren't supported yet.
- MesonProject now monitors all meson.build files and regenerates the project if they change.
- Made extensive use of cancellable to make the server responsive on shutdown requests.
- References that belong to a file (`*.vapi`) generated by another target are now resolved to references in the specific file in that target.

Finally, the reorganization of the project backend will make it much easier to implement new project backends like CMake or Autotools.

**Comment by Prince781** [comment]
@benwaffle @gavr123456789 if you could just look these changes over very quickly—you don't even have to comment much really—that would be all. I've already tested this branch but maybe I've overlooked something.

**Comment by benwaffle** [comment]
I'd prefer to do a real code review but i can't guarantee the timing of that

**Comment by Prince781** [comment]
@benwaffle fair enough. I know you're probably busy with things right now, and I've already tested this a lot (and I got Gavr to test it too). So I'll merge it and take responsibility for if something goes wrong.

File: src/main.vala [hunk 1]
```diff
@@ -256,288 +245,64 @@ class Vls.Server : Object {
...
+        init_params = Util.parse_variant<InitializeParams> (@params);
 
-        var root_dir = File.new_for_uri (init_params.rootUri);
+        File root_dir;
+        if (init_params.rootUri != null)
+            root_dir = File.new_for_uri (init_params.rootUri);
+        else if (init_params.rootPath != null)
+            root_dir = File.new_for_path (init_params.rootPath);
+        else
+            root_dir = File.new_for_path (Environment.get_current_dir ());
```
Comment (benwaffle) [inline src/main.vala:263]: what's the CWD of the server?

**Pull request merged**